README.md
# HaskellTreeDSL

A minimal [BASIC](https://en.wikipedia.org/wiki/BASIC)-like [Domain Specific Language](https://en.wikipedia.org/wiki/Domain-specific_language) (DSL), implemented in [Haskell](https://www.haskell.org) 🧐.

This project is a demo, with 'minimum viable' BASIC language features largely implemented through AI code generation.

![Festive tree](.res/tree-ghci-50.png)

The BASIC code, for the tree _outline_ (without decorations), looks like this:
```basic
10 FOR i = 1 TO 12
20 LET spaces = 12 - i
30 LET tree = i * 2 - 1
40 FOR j = 1 TO spaces
45 COLOR "Green"
50 PRINT " ";
60 NEXT j
70 FOR k = 1 TO tree
80 PRINT "/";
90 NEXT k
100 PRINT
110 NEXT i
125 COLOR "Brown"
120 PRINT "         ||||"
125 COLOR "Red"
130 PRINT "       ________"
```

## Background

ChatGPT generated the BASIC outline for the tree, without decorations, and the fundamentals of the Haskell parser.

To improve the accuracy of the generated code, a custom GPT was used, [Haskell Parser Helper GPT](https://chat.openai.com/g/g-mwn9vJgLk-haskell-parser-helper), with key help files from [Megaparsec](https://github.com/mrkkrp/megaparsec) loaded as knowledge. This did seem to improve the generated code accuracy to a degree, but further refinement is needed. The parser is not entirely AI generated, it needed a little tweaking to fix a couple of end of line / keyword handling issues that crept in.


## devcontainer
This project 'should' launch in a devcontainer: https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers

## run
build and run as a binary:
```bash
stack run
```

## to debug / develop
run in ghci:
```bash
stack ghci

ghci> SamplePrograms.main
```
`SamplePrograms.hs` contains other programs to run, with a little tweaking.

## test cases
run the test cases:
```bash
stack ghci --test

(option 2)

ghci> Main.main
```


---
## Additional notes:

#### Testing
List the tests:
```bash
stack test --test-arguments "--dry-run"
```

Run a specific test:
```bash
stack test --test-arguments "-m \"parses valid LET statement"\"

stack test --test-arguments "-m \"parses valid LET statement with arithmetic"\"

stack test --test-arguments "-m \"FOR NEXT program tests/parses valid FOR statement in program, with PRINT"\"

stack test --test-arguments "-m \"FOR NEXT tests"\"
```

#### traceShow
Add `traceShow` within parsers to log the parser state:
```haskell
  -- Debugging prints
  let debugInfo = (ln, varName, startExpr, endExpr)
  traceShow debugInfo $ return ()
```


CHANGELOG.md
# Changelog for `HaskellTreeDSL`

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to the
[Haskell Package Versioning Policy](https://pvp.haskell.org/).

## Unreleased

## 0.1.0.0 - YYYY-MM-DD


test/Spec.hs
module Main (main) where

import qualified Parse.ForSpec (allTests)
import qualified Parse.PrintSpec (allTests)
import qualified Parse.IfSpec (allTests)
import qualified SimpleProgramSpec
import qualified ExprSpec (allTests)
import Test.Hspec


main :: IO ()
main = hspec $ do
--  ParseDefsSpec.allTests
  ExprSpec.allTests
  Parse.PrintSpec.allTests
  Parse.ForSpec.allTests
  Parse.IfSpec.allTests

  SimpleProgramSpec.allTests



test/ExprSpec.hs
module ExprSpec (main, allTests) where

import Test.Hspec
import Text.Megaparsec
import Test.Hspec.Megaparsec
-- import ParseDefs -- Your module with the parsers
import Parser.Expr
import Parser.Parser

termParserTests :: Spec
termParserTests =
  describe "term parser" $ do

    describe "multiplication" $ do

      it "parses simple multiplication, no spaces" $ do
        let exptected = Mult (ConstInt 3) (ConstInt 4)
        parse term "" "3*4" `shouldParse` exptected

      it "parses simple multiplication" $ do
        let exptected = Mult (ConstInt 3) (ConstInt 4)
        parse term "" "3 * 4" `shouldParse` exptected

      it "parses simple multiplication, no spaces" $ do
        let exptected = Mult (ConstInt 3) (ConstInt 4)
        parse term "" "3* 4" `shouldParse` exptected

      it "parses simple multiplication, no spaces" $ do
        let exptected = Mult (ConstInt 3) (ConstInt 4)
        parse term "" "3 *4" `shouldParse` exptected

      it "parses simple multiplication, no spaces, three parameters" $ do
        let exptected = Mult (Mult (ConstInt 3) (ConstInt 4)) (ConstInt 5)
        parse term "" "3*4*5" `shouldParse` exptected

    describe "divsion" $ do

      it "12 divided by 4, no spaces" $ do
        let exptected = Div (ConstInt 12) (ConstInt 4)
        parse term "" "12/4" `shouldParse` exptected


      it "12 divided by i, no spaces" $ do
        let exptected = Div (ConstInt 12) (Var "i")
        parse term "" "12/i" `shouldParse` exptected



arithParserTests :: Spec
arithParserTests =
  describe "term parser" $ do

    describe "addition" $ do
      it "12 plus 4" $ do
        let exptected = Add (ConstInt 12) (ConstInt 4)
        parse arithExpr "" "12 + 4" `shouldParse` exptected

      it "12 plus 4, no spaces" $ do
        let exptected = Add (ConstInt 12) (ConstInt 4)
        parse arithExpr "" "12+4" `shouldParse` exptected

      it "12 plus 4 plus 3" $ do
        let exptected = Add (Add (ConstInt 12) (ConstInt 4)) (ConstInt 3)
        parse arithExpr "" "12 + 4 + 3" `shouldParse` exptected


    describe "subtraction" $ do
      it "12 minus 4" $ do
        let exptected = Sub (ConstInt 12) (ConstInt 4)
        parse arithExpr "" "12 - 4" `shouldParse` exptected

      it "12 minus 4, no spaces" $ do
        let exptected = Sub (ConstInt 12) (ConstInt 4)
        parse arithExpr "" "12-4" `shouldParse` exptected

      it "12 minus 4 minus 3" $ do
        let exptected = Sub (Sub (ConstInt 12) (ConstInt 4)) (ConstInt 3)
        parse arithExpr "" "12 - 4 - 3" `shouldParse` exptected


opPrecParserTests :: Spec
opPrecParserTests = do
  describe "op prec parser" $ do

    describe "mixed" $ do
      it "12 + 4 * 3" $ do
        let exptected = Add (ConstInt 12) (Mult (ConstInt 4) (ConstInt 3))
        parse arithExpr "" "12 + 4 * 3" `shouldParse` exptected




varExprTests :: Spec
varExprTests = 
  describe "varExpr Parser" $ do
    it "parses valid variable expressions" $ do
      parse varExpr "" "variable " `shouldParse` Var "variable"
    
    it "fails to parse invalid variable expressions" $ do
      parse varExpr "" `shouldFailOn` "123invalid "

    it "fails to parse invalid variable expressions" $ do
      parse varExpr "" `shouldFailOn` "123invalid%"

    it "parses variable expressions without trailing space" $ do
      parse varExpr "" "variable123" `shouldParse` Var "variable123"


stringExprTests :: Spec
stringExprTests = describe "stringExpr and stringLiteral Parser" $ do

  describe "stringExpr and stringLiteral Parser" $ do
    it "parses valid string expressions" $ do
      parse stringExpr "" "\"Hello, World!\"" `shouldParse` ConstStr "Hello, World!"

    it "parses empty string expressions" $ do
      parse stringExpr "" "\"\"" `shouldParse` ConstStr ""

    it "fails to parse invalid string expressions" $ do
      parse stringExpr "" `shouldFailOn` "Hello, World!"

    it "fails to parse incomplete string expressions" $ do
      parse stringExpr "" `shouldFailOn` "\"Hello, World"



numberExprTests :: Spec
numberExprTests = describe "numberExprTest Parser" $ do

  describe "number Parser" $ do
    it "parses positive integers followed by space" $ do
      parse number "" "123 " `shouldParse` 123

    it "parses positive integers without trailing space" $ do
      parse number "" "123" `shouldParse` 123

    it "fails to parse non-integer input" $ do
      parse number "" `shouldFailOn` "abc"

  describe "unaryExpr Parser" $ do
    it "parses negative integers" $ do
      parse unaryExpr "" "-123 " `shouldParse` Negate (ConstInt 123)


    it "fails to parse negative integers with space" $ do
      parse unaryExpr "" `shouldFailOn` "- 123"


    it "parses negative variables" $ do
      parse unaryExpr "" "-variable " `shouldParse` Negate (Var "variable")


    it "fails to parse non-unary expressions" $ do
      parse unaryExpr "" `shouldFailOn` "123"


constExprTests :: Spec
constExprTests = describe "constExprTests parser" $ do

  describe "ConstInt Parser" $ do
    it "parses positive integer, space after" $ do
      parse constExpr "" "123 " `shouldParse` ConstInt 123

  describe "ConstInt Parser" $ do
    it "parses positive integer, no spaces" $ do
      parse constExpr "" "890" `shouldParse` ConstInt 890

  describe "ConstInt Parser" $ do
    it "parses positive integer, space before" $ do
      parse constExpr "" `shouldFailOn` " 890"

  describe "ConstInt Parser" $ do
    it "parses positive integer, space before and after" $ do
      parse constExpr "" `shouldFailOn` " 899 " 



conditionTests :: Spec
conditionTests = describe "expr parser - conditions" $ do

    it "parses a = 1" $ do
      parse expr "" "a = 1" `shouldParse` Eq (Var "a") (ConstInt 1)

    it "parses a=1" $ do
      parse expr "" "a=1" `shouldParse` Eq (Var "a") (ConstInt 1)


exprTests :: Spec
exprTests = describe "expr parser" $ do

    it "parses positive integers followed by space" $ do
      parse expr "" "123 " `shouldParse` ConstInt 123

    it "parses negative variables" $ do
      parse expr "" "-variable " `shouldParse` Negate (Var "variable")

    it "fails to parse negative integers with space" $ do
      parse expr "" `shouldFailOn` "- 123"

    it "parses negative variables" $ do
      parse expr "" "-variable " `shouldParse` Negate (Var "variable")

    it "parses valid string expressions" $ do
      parse expr "" "\"Hello, World!\"" `shouldParse` ConstStr "Hello, World!"


andTests :: Spec
andTests = describe "expr AND parser" $ do

    it "AND const const" $ do
      parse expr "" "123 AND 456" `shouldParse` And (ConstInt 123) (ConstInt 456)

    it "AND var=0 const" $ do
      let expected = And (Eq (Var "i") (ConstInt 0)) (ConstInt 456)
      parse expr "" "i = 0 AND 456" `shouldParse` expected

    it "AND var=0 var=1" $ do
      let expected = And (Eq (Var "i") (ConstInt 0)) (Eq (Var "o") (ConstInt 10)) 
      parse expr "" "i = 0 AND o = 10" `shouldParse` expected



allTests :: Spec
allTests = do
  numberExprTests
  constExprTests
  termParserTests
  arithParserTests
  opPrecParserTests
  varExprTests
  stringExprTests  
  exprTests
  conditionTests
  andTests


main :: IO ()
main = hspec $ do 
  allTests
  --andTests


test/Parse/IfSpec.hs
module Parse.IfSpec (main,allTests) where

import Test.Hspec ( hspec, describe, it, Spec )
import Text.Megaparsec
import Test.Hspec.Megaparsec ( shouldParse )
import Parser.Expr
import Parser.Stmt
import Parser.Parser

ifThenTests :: Spec
ifThenTests =
  describe "IF THEN END" $ do
    it "parses empty IF ELSE END" $ do
      let input = unlines 
            [ 
              "30 IF a = 0 THEN",
              "50 ELSE",
              "60 END"
            ]

      let expected = If (LineNumber 30) (Eq (Var "a") (ConstInt 0)) [] (Just [])
      parse ifStmt "" input `shouldParse` expected

    it "parses IF END no ELSE" $ do
      let input = unlines 
            [ 
              "30 IF a = 0 THEN",
              "60 END"
            ]

      let expected = If (LineNumber 30) (Eq (Var "a") (ConstInt 0)) [] Nothing
      parse ifStmt "" input `shouldParse` expected


ifThenProgramTests :: Spec
ifThenProgramTests = describe "IF THEN tests" $ do
  it "parses valid IF THEN END statements in program" $ do
    let input = unlines 
            [ 
              "20 LET a = 0",
              "30 IF a = 0 THEN",
              "50 REM PRINT \"x\";",
              "60 END"
            ]
    let expected =
          [Let (LineNumber 20) "a" (ConstInt 0), 
          If (LineNumber 30) (Eq (Var "a") (ConstInt 0)) [Rem (LineNumber 50) "PRINT \"x\";"] Nothing
          ] 
    parse program "" input `shouldParse` expected





  it "parses valid PRINT statements in program with semicolon" $ do
    let input = unlines 
            [ 
              "20 LET a = 0",
              "30 IF a = 0 THEN",
              "50 PRINT \"x\";",
              "60 END"
            ]
    let expected0 =
           [Let (LineNumber 20) "a" (ConstInt 0), 
           If (LineNumber 30) (Eq (Var "a") (ConstInt 0)) [Print (LineNumber 50) (ConstStr  "x") False] Nothing]
    parse program "" input `shouldParse` expected0


ifThenProgramElseTests :: Spec
ifThenProgramElseTests = describe "IF THEN ELSE tests" $ do


  it "parses valid IF THEN ELSE END statements in program" $ do
    let input = unlines 
            [ 
              "20 LET a = 0",
              "30 IF a = 0 THEN",
              "50 REM In body",
              "60 ELSE",
--              "70 REM In else",
              "70 PRINT",
              "80 END"
            ]
    let expected =
          [ Let (LineNumber 20) "a" (ConstInt 0), 
            If (LineNumber 30) (Eq (Var "a") (ConstInt 0)) [Rem (LineNumber 50) "In body"] (Just [Print (LineNumber 70) (ConstStr "") True])
          ] 
    putStrLn input
    parse program "" input `shouldParse` expected


  it "parses valid IF THEN ELSE END statements in program, with code following" $ do
    let input = unlines 
            [ 
              "20 LET a = 0",
              "30 IF a = 0 THEN",
              "50 REM In body",
              "60 ELSE",
              "70 PRINT",
              "80 END",
              "90 PRINT \"After\""
            ]
    let expected =
          [ Let (LineNumber 20) "a" (ConstInt 0), 
            If (LineNumber 30) (Eq (Var "a") (ConstInt 0)) [Rem (LineNumber 50) "In body"] (Just [Print (LineNumber 70) (ConstStr "") True]),
            Print (LineNumber 90) (ConstStr "After") True
          ] 
    putStrLn input
    parse program "" input `shouldParse` expected

  it "parses valid IF THEN ELSE END, multiple statements in ELSE" $ do
    let input = unlines 
            [ 
              "20 LET a = 0",
              "30 IF a = 0 THEN",
              "50 REM In body",
              "60 ELSE",
              "70 PRINT",
              "75 PRINT a",
              "80 END",
              "90 PRINT \"After\""
            ]
    let expected =
          [ Let (LineNumber 20) "a" (ConstInt 0), 
            If (LineNumber 30) (Eq (Var "a") (ConstInt 0)) 
              [Rem (LineNumber 50) "In body"] 
              (Just [Print (LineNumber 70) (ConstStr "") True,
                Print (LineNumber 75) (Var "a") True
              ]),
            Print (LineNumber 90) (ConstStr "After") True
          ] 
    putStrLn input
    parse program "" input `shouldParse` expected



  it "parses valid IF THEN ELSE END, nested IF THEN END in ELSE" $ do
    let input = unlines 
            [ 
              "20 LET a = 0",
              "30 IF a = 0 THEN",
              "50 REM In body",
              "60 ELSE",
              "65 IF a = 1 THEN",
              "70 PRINT \"a is 1\"",
              "75 END",
              "80 END",
              "90 PRINT \"After\""
            ]
    let expected =
          [ Let (LineNumber 20) "a" (ConstInt 0), 
            If (LineNumber 30) (Eq (Var "a") (ConstInt 0)) 
              [Rem (LineNumber 50) "In body"] 
              (Just [If (LineNumber 65) (Eq (Var "a") (ConstInt 1)) 
                [ Print (LineNumber 70) (ConstStr "a is 1") True] Nothing]), 
            Print (LineNumber 90) (ConstStr "After") True]

    putStrLn input
    parse program "" input `shouldParse` expected



allTests :: Spec
allTests = do
  ifThenTests
  ifThenProgramTests
  ifThenProgramElseTests


main :: IO ()
main = hspec $ do
  allTests  
  -- ifThenTests
  -- ifThenProgramTests
  -- ifThenProgramElseTests


src/Colors.hs

{-# LANGUAGE OverloadedStrings #-}

module Colors (main, resetColor, textColor) where

import System.Console.ANSI

resetColor :: IO ()
resetColor = setSGR [Reset]

-- Black	 
-- Red	 
-- Green	 
-- Yellow	 
-- Blue	 
-- Magenta	 
-- Cyan	 
-- White
-- https://hackage.haskell.org/package/ansi-terminal-types-0.11.5/docs/System-Console-ANSI-Types.html

textColor :: String -> IO ()
textColor color = 
  -- map color to ANSI color
  case color of
    "Red" -> setSGR [SetColor Foreground Vivid Red]
    "Blue" -> setSGR [SetColor Foreground Vivid Blue]
    "Yellow" -> setSGR [SetColor Foreground Vivid Yellow]
    "Brown" -> setSGR [SetColor Foreground Dull Red]
    "Green" -> setSGR [SetColor Foreground Vivid Green]
    "Magenta" -> setSGR [SetColor Foreground Vivid Magenta]
    "Cyan" -> setSGR [SetColor Foreground Vivid Cyan]
    _ -> setSGR [Reset]  -- Reset to default


main :: IO ()
main = do
  setSGR [SetColor Foreground Vivid Red]
  putStrLn "This will be red"
  setSGR [Reset]  -- Reset to default

  setSGR [SetColor Foreground Vivid Blue]
  putStr "This will be Blue "
  setSGR [SetColor Foreground Vivid Yellow]
  putStrLn " and this will be Yellow "

  setSGR [Reset]  -- Reset to default

src/Interpreter.hs
{-# LANGUAGE OverloadedStrings #-}

{-# OPTIONS_GHC -Wno-unused-matches #-}

module Interpreter (runProgram, runProgram') where

import Parser.Stmt (Stmt(..), Expr(..))
import Control.Monad.State
import Data.Map (Map)
import qualified Data.Map as Map
import Colors (resetColor, textColor)

data Value = IntVal Int | StrVal String
    deriving (Show, Eq)


displayValue :: Value -> String
displayValue (IntVal i) = show i
displayValue (StrVal s) = s


type InterpreterState = Map String Value


interpret :: [Stmt] -> StateT InterpreterState IO ()
interpret = mapM_ interpretStmt

interpretStmt :: Stmt -> StateT InterpreterState IO ()
interpretStmt (Rem ln var) = do
    -- liftIO $ putStrLn $ "REM> " ++ var
    return ()

interpretStmt (Let ln var expr) = do
    value <- evaluateExpr expr
    modify (Map.insert var value)

interpretStmt (Print ln expr addNewLine) = do
    value <- evaluateExpr expr
    liftIO $ putStr $ displayValue value
    when addNewLine $ do
        liftIO $ putStrLn ""

        liftIO $ resetColor
        liftIO $ putStr "BASIC> "
        currentColor <- gets (Map.lookup "currentColor")
        case currentColor of
            Just (StrVal color) -> liftIO $ textColor color
            _ -> return ()    


interpretStmt (For ln loopVar startExpr endExpr loopStmts) = do
    startVal <- evaluateExpr startExpr
    endVal <- evaluateExpr endExpr
    case (startVal, endVal) of
        (IntVal s, IntVal e) -> forLoop loopVar s e loopStmts
        _ -> error "Type mismatch: For loop expects integer values"
  where
    forLoop var currentVal endVal stmts
        | currentVal > endVal = return ()  -- Exit condition
        | otherwise = do
            modify (Map.insert var (IntVal currentVal))  -- Update loop variable
            mapM_ interpretStmt stmts                    -- Execute loop body
            forLoop var (currentVal + 1) endVal stmts    -- Recursive call for next iteration

interpretStmt (If ln condExpr stmts elseStmts) = do
    condVal <- evaluateExpr condExpr
    case condVal of
        IntVal n ->
            if n /= 0   -- Non-zero integers are true
            then mapM_ interpretStmt stmts  
            else maybe (return ()) (mapM_ interpretStmt) elseStmts 

        StrVal _ -> error "Type mismatch: If condition expects an integer value"
        -- Or, if you prefer, treat any non-empty string as true and empty string as false
        -- StrVal s -> when (not $ null s) $ mapM_ interpretStmt stmts

interpretStmt (Color _ color) = do
    -- liftIO $ putStrLn $ "COLOR> " ++ color
    modify (Map.insert "currentColor" (StrVal color))
    liftIO $ textColor color


-- interpretStmt s = do
--     _ <- error $ "Not implemented " ++ show s
--     return ()


{-
Expression evaluation
-}

evaluateExpr :: Expr -> StateT InterpreterState IO Value
evaluateExpr (Var x) = gets (Map.! x)
evaluateExpr (ConstInt n) = return $ IntVal n
evaluateExpr (ConstStr str) = return $ StrVal str

evaluateExpr (Sub e1 e2) = do
    v1 <- evaluateExpr e1
    v2 <- evaluateExpr e2
    case (v1, v2) of
        (IntVal i1, IntVal i2) -> return $ IntVal (i1 - i2)  -- Adding integers
        (StrVal _, StrVal _) -> return $ IntVal 0 -- Quite BASIC return NaN
        _ -> error "Type mismatch: Cannot add integer and string"  -- Type mismatch error

evaluateExpr (Add e1 e2) = do
    v1 <- evaluateExpr e1
    v2 <- evaluateExpr e2
    case (v1, v2) of
        (IntVal i1, IntVal i2) -> return $ IntVal (i1 + i2)  -- Adding integers
        (StrVal s1, StrVal s2) -> return $ StrVal (s1 ++ s2)  -- Concatenating strings
        _ -> error "Type mismatch: Cannot add integer and string"  -- Type mismatch error

evaluateExpr (Mult e1 e2) = do
    v1 <- evaluateExpr e1
    v2 <- evaluateExpr e2
    case (v1, v2) of
        (IntVal i1, IntVal i2) -> return $ IntVal (i1 * i2)  -- Adding integers
        _ -> error "Type mismatch: Cannot multiply"  -- Type mismatch error

evaluateExpr (Div e1 e2) = do
    v1 <- evaluateExpr e1
    v2 <- evaluateExpr e2
    case (v1, v2) of
        (IntVal i1, IntVal i2) -> return $ IntVal (div i1 i2)  -- Adding integers
        _ -> error "Type mismatch: Cannot divide"  -- Type mismatch error




-- evaluateExpr (Sub e1 e2) = do
--     v1 <- evaluateExpr e1
--     v2 <- evaluateExpr e2
--     case (v1, v2) of
--         (IntVal i1, IntVal i2) -> return $ IntVal (i1 - i2)  -- Sub integers
--         (StrVal s1, StrVal s2) -> return $ StrVal (s1 ++ s2)  -- Concatenating strings
--         _ -> error "Type mismatch: Cannot add integer and string"  -- Type mismatch error

evaluateExpr (Eq e1 e2) = do
    v1 <- evaluateExpr e1
    v2 <- evaluateExpr e2
    --return (if v1 == v2 then 1 else 0)
    case (v1, v2) of
        (IntVal i1, IntVal i2) -> return $ IntVal (if i1 == i2 then 1 else 0)
        (StrVal s1, StrVal s2) -> return $ IntVal (if s1 == s2 then 1 else 0)
        _ -> return $ IntVal 0


evaluateExpr (Negate e1) = do
    v1 <- evaluateExpr e1
    --return (if v1 == v2 then 1 else 0)
    case v1 of
        (IntVal i1) -> return $ IntVal (-i1)
        _ -> error "Cannot negate string"


evaluateExpr (And e1 e2) = do
    v1 <- evaluateExpr e1
    v2 <- evaluateExpr e2
    return $ v1 `andOp` v2

andOp :: Value -> Value -> Value
andOp (IntVal a) (IntVal b) = IntVal $ if a /= 0 && b /= 0 then 1 else 0
andOp _ _ = error "Type mismatch: AND operator expects integer values"


runProgram :: [Stmt] -> IO InterpreterState
runProgram stmts = do
    resetColor
    putStrLn "\nRunning"
    result <- execStateT (interpret stmts) initialState
    resetColor
    putStrLn ""
    return result
  where
    initialState = Map.empty -- or any default initial state you prefer


-- run without color reset
runProgram' :: [Stmt] -> IO InterpreterState
runProgram' stmts = execStateT (interpret stmts) initialState
  where
    initialState = Map.empty -- or any default initial state you prefer





src/Parser/Expr.hs
{-# LANGUAGE OverloadedStrings #-}

module Parser.Expr
  ( expr,
    Expr (..),
    identifier,
    varExpr,
    constExpr,
    stringLiteral,
    -- test
    unaryExpr,
    number,
    stringExpr,
    arithExpr,
    term
  )
where

import Control.Monad.Combinators
import Parser.Common (Parser)
import Text.Megaparsec
import Text.Megaparsec.Char
  ( alphaNumChar,
    char,
    letterChar,
    space,
    space1,
    string,
  )
import qualified Text.Megaparsec.Char.Lexer as L

-- import Text.Megaparsec.Debug (dbg)
-- import Debug.Trace (traceShow, traceShowId)

-- Parses an integer
number :: Parser Int
number = L.decimal -- <* space

-- Parses specific keyword (e.g. variable name)
identifier :: Parser String
identifier = (:) <$> letterChar <*> many alphaNumChar

-- Expressions

data Expr
  = Var String
  | ConstInt Int
  | ConstStr String
  | Add Expr Expr
  | Sub Expr Expr
  | Mult Expr Expr
  | Div Expr Expr
  | Eq Expr Expr -- =
  | Negate Expr
  | And Expr Expr
  deriving (Show, Eq)

unaryExpr :: Parser Expr
unaryExpr = do
  _ <- char '-'
  expr' <- constExpr <|> varExpr -- Use the appropriate sub-parser here
  return $ Negate expr'

constExpr :: Parser Expr
constExpr = ConstInt <$> number

varExpr :: Parser Expr
varExpr = Var <$> identifier

stringExpr :: Parser Expr
stringExpr = ConstStr <$> stringLiteral

stringLiteral :: Parser String
stringLiteral = char '"' >> manyTill L.charLiteral (char '"')

factor :: Parser Expr
factor = try unaryExpr <|> try constExpr <|> try stringExpr <|> varExpr -- <|> parenthesizedExpr

operatorParser :: Parser (Expr -> Expr -> Expr)
operatorParser = do
  space -- Consume leading spaces
  op <- Mult <$ char '*' <|> Div <$ char '/' -- Parse the operator
  space -- Consume trailing spaces
  return op

arithOperatorParser :: Parser (Expr -> Expr -> Expr)
arithOperatorParser = do
  space -- Consume leading spaces
  op <- Add <$ char '+' <|> Sub <$ char '-' -- Parse the operator
  space -- Consume trailing spaces
  return op

term :: Parser Expr
term = do
  e1 <- factor
  more <- many $ try $ (,) <$> operatorParser <*> factor
  return $ foldl (\acc (op, e) -> op acc e) e1 more

arithExpr :: Parser Expr
arithExpr = do
  e1 <- term
  more <- many $ try $ (,) <$> arithOperatorParser <*> term
  return $ foldl (\acc (op, e) -> op acc e) e1 more

keywordAnd :: Parser String
keywordAnd = space1 *> string "AND" <* space1

andExpr :: Parser Expr
andExpr = do
  left <- primaryExpr
  _ <- keywordAnd
  And left <$> primaryExpr

eqExpr :: Parser Expr
eqExpr = do
  e1 <- varExpr <|> constExpr
  _ <- optional space
  _ <- string "=" <* space
  e2 <- varExpr <|> constExpr
  return $ Eq e1 e2

primaryExpr :: Parser Expr
primaryExpr = try eqExpr <|> arithExpr

expr :: Parser Expr
expr = try andExpr <|> primaryExpr


src/SamplePrograms.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE QuasiQuotes #-}
{-# OPTIONS_GHC -Wno-unused-imports #-}
{-# OPTIONS_GHC -Wno-unused-top-binds #-}

module SamplePrograms (main, treeDemoA) where

import Interpreter (runProgram)
import Text.RawString.QQ (r)
import Text.Megaparsec ( errorBundlePretty )
import Parser.Parser (parseBasicProgram')


treeDemoA :: String
treeDemoA = unlines 
  [
    "10 PRINT",
    "20 FOR i = 1 TO 12",
    "25 REM place the star",
    "30 IF i = 1 THEN",
    "40 COLOR \"Yellow\"",
    "45 FOR s = 1 TO 11",
    "46 PRINT \" \";",
    "47 NEXT s",
    "48 PRINT \"*\"",
    "60 END",
    "65 REM draw the tree body",
    "70 LET spaces = 12 - i",
    "80 LET tree = i * 2 - 1",
    "90 FOR j = 1 TO spaces",
    "110 PRINT \" \";",
    "120 NEXT j",
    "130 FOR k = 1 TO tree",
    "140 REM place decorations or tree body",
--
    "150 IF i = 2 AND k = 1 THEN",
    "160 COLOR \"Red\"",
    "170 PRINT \"O\";",
    "180 ELSE",
    "181 IF i = 5 AND k = 7 THEN",  -- First additional condition
    "182 COLOR \"Magenta\"",
    "183 PRINT \"O\";",
    "184 ELSE",
    "185 IF i = 7 AND k = 6 THEN",  -- Second additional condition
    "186 COLOR \"Red\"",
    "187 PRINT \"O\";",
    "188 ELSE",
    "189 IF i = 9 AND k = 12 THEN",  -- Third additional condition (original condition repeated)
    "190 COLOR \"Cyan\"",
    "191 PRINT \"O\";",
    "192 ELSE",
    "193 IF i = 11 AND k = 9 THEN",  -- Fourth additional condition
    "194 COLOR \"Yellow\"",
    "195 PRINT \"O\";",
    "196 ELSE",
    "197 COLOR \"Green\"",
    "198 PRINT \"/\";",
    "199 END",
    "200 END",
    "201 END",
    "202 END",
    "203 END",
    "204 REM end of decorations or tree body",
--
    "220 NEXT k",
    "230 PRINT",
    "240 NEXT i",
    "250 COLOR \"Brown\"",
    "260 PRINT \"         ||||\"",
    "270 COLOR \"Red\"",
    "280 PRINT \"       ________\""
  ]



treeDemoA'' :: String
treeDemoA'' =
  [r|
10 PRINT
20 FOR i = 1 TO 12
25 REM place the star
30 IF i = 1 THEN
40 COLOR "Yellow"
45 FOR s = 1 TO 11
46 PRINT " ";
47 NEXT s
48 PRINT "*"
60 END
65 REM draw the tree body
70 LET spaces = 12 - i
80 LET tree = i * 2 - 1
90 FOR j = 1 TO spaces
110 PRINT " ";
120 NEXT j
130 FOR k = 1 TO tree
140 REM place decorations or tree body
150 IF i = 2 AND k = 1 THEN
160 COLOR "Red"
170 PRINT "O";
180 ELSE
181 IF i = 5 AND k = 1 THEN
182 COLOR "Red"
183 PRINT "O";
184 END
190 COLOR "Green"
200 PRINT "/";
210 END
220 NEXT k
230 PRINT
240 NEXT i
250 COLOR "Brown"
260 PRINT "         ||||"
270 COLOR "Red"
280 PRINT "       ________"
|]


treeDemoA' :: String
treeDemoA' =
  [r|
5 PRINT
10 FOR i = 1 TO 12
15 IF i = 1 THEN
16 COLOR "Yellow"
17 PRINT "           *"
18 END
20 LET spaces = 12 - i
30 LET tree = i * 2 - 1
40 FOR j = 1 TO spaces
45 COLOR "Green"
50 PRINT " ";
60 NEXT j
70 FOR k = 1 TO tree
80 PRINT "/";
90 NEXT k
100 PRINT
110 NEXT i
125 COLOR "Brown"
120 PRINT "         ||||"
125 COLOR "Red"
130 PRINT "       ________"
|]


treeDemoB :: String
treeDemoB =
  [r|
5 PRINT "."
10 FOR i = 1 TO 8
20 LET spaces = 8 - i
30 LET stars = i * 2 - 1
40 FOR j = 1 TO spaces
50 PRINT " ";
60 NEXT j
70 FOR k = 1 TO stars
80 PRINT "/";
90 NEXT k
100 PRINT
110 NEXT i
120 PRINT "      ||||"
130 PRINT "    ________"
|]


longerTestProg :: String
longerTestProg =
  [r|
5 REM test
10 LET z = 6
20 FOR i = 1 TO 10
30 PRINT i
40 IF i = z THEN
50 PRINT "Boom!"
60 END
70 NEXT i
80 LET x = z + 1
90 PRINT "Expecting 7"
100 PRINT x
110 PRINT 5 - 2
120 LET a = -1
130 PRINT a
135 PRINT "Should print 6 >"
140 PRINT 7 + a
145 PRINT "Should print 8 >"
150 PRINT 2 * 4
155 PRINT "Should print 3 >"
160 PRINT 9 / 3
|]


ifElseTestProg :: String
ifElseTestProg =
  [r|
10 REM test
20 FOR i = 1 TO 10
25 PRINT i
30 IF i = 5 THEN
40 PRINT "Five"
50 ELSE
60 PRINT "Not five"
80 END
90 NEXT i
|]

ifElseAndTestProg :: String
ifElseAndTestProg =
  [r|
10 REM test
20 FOR i = 1 TO 3
24 FOR o = 2 TO 5
25 PRINT i
26 PRINT o
30 IF i = 2 AND o = 5 THEN
40 PRINT "2 AND 5"
50 ELSE
60 PRINT "other"
80 END
85 NEXT o
90 NEXT i
|]


mainEProgram :: String
mainEProgram = [r|
10 LET z = 7
20 FOR i = 1 TO 10
35 PRINT i
40 IF i = z THEN
50 PRINT "Boom!"
60 END
70 NEXT i
|]




main :: IO ()
main = do
  -- let input = longerTestProg
  let input = treeDemoA
  -- let input = ifElseTestProg
  --let input = ifElseAndTestProg
  putStrLn "Input program\n"
  putStrLn input
  putStrLn ""
  case parseBasicProgram' input of
    Left bundle -> putStr (errorBundlePretty bundle) -- Print parsing errors
    Right ast -> do
        putStrLn "Abstract syntax tree\n"
        print ast -- Print the abstract syntax tree (AST)
        _ <- runProgram ast
        return ()